File: typescript/phone-number/_iterations/1/phone-number.ts
export function clean(input: string): string {
  if (/[A-Za-z]/.test(input)) {
    throw new Error('Letters not permitted');
  } else if (!/^\s*(\+?\d)?\s*\(?\s*\d+\s*\)?[\s-.]*\d+[\s-.]*\d+[\s-.]*\d+\s*$/.test(input)) {
    throw new Error('Punctuations not permitted');
  }

  let numbers: string = input.replaceAll(/[^\d]+/g, '');

  if (numbers.length < 10) {
    throw new Error('Incorrect number of digits');
  } else if (numbers.length > 11) {
    throw new Error('More than 11 digits');
  }

  if (numbers.length === 11) {
    if (numbers[0] !== '1') {
      throw new Error('11 digits must start with 1');
    }
    numbers = numbers.substring(1);
  }

  const areaCode = numbers[0];
  if (areaCode === '0') {
    throw new Error('Area code cannot start with zero');
  } else if (areaCode === '1') {
    throw new Error('Area code cannot start with one');
  }
  
  const exchangeCode = numbers[3];
  if (exchangeCode === '0') {
    throw new Error('Exchange code cannot start with zero');
  } else if (exchangeCode === '1') {
    throw new Error('Exchange code cannot start with one');
  }

  return numbers;
}

File: java/phone-number/_iterations/1/src/main/java/PhoneNumber.java
import java.util.regex.Pattern;

public class PhoneNumber {
    private static final Pattern PUNCTUATIONS = Pattern.compile("[\\s.()-]");
    private static final Pattern LETTER = Pattern.compile("[a-zA-Z]");
    private static final Pattern NON_NUMBER = Pattern.compile("\\D");

    private String cleanedNumber;

    PhoneNumber(String inputNumber) {
        this.cleanedNumber = cleanup(inputNumber);
    }

    public String getNumber() {
        return cleanedNumber;
    }

    private static String cleanup(String inputNumber) {
        require(inputNumber != null, "number cannot be null");
        String cleanedNumber = PUNCTUATIONS.matcher(inputNumber).replaceAll("");
        if (cleanedNumber.startsWith("+")) {
            cleanedNumber = cleanedNumber.substring(1);
        }
        require(!LETTER.matcher(cleanedNumber).find(), "letters not permitted");
        require(!NON_NUMBER.matcher(cleanedNumber).find(), "punctuations not permitted");
        require(cleanedNumber.length() <= 11, "more than 11 digits");
        if (cleanedNumber.length() == 11) {
            require(cleanedNumber.startsWith("1"), "11 digits must start with 1");
            cleanedNumber = cleanedNumber.substring(1);
        }
        require(cleanedNumber.length() == 10, "incorrect number of digits");
        char areaCodeStart = cleanedNumber.charAt(0);
        require(areaCodeStart != '0', "area code cannot start with zero");
        require(areaCodeStart != '1', "area code cannot start with one");
        char exchangeCodeStart = cleanedNumber.charAt(3);
        require(exchangeCodeStart != '0', "exchange code cannot start with zero");
        require(exchangeCodeStart != '1', "exchange code cannot start with one");

        return cleanedNumber;
    }

    private static void require(Boolean condition, String message) {
        if (!condition) {
            throw new IllegalArgumentException(message);
        }
    }
}
